stan/math/prim/mat/prob/std_binormal_lcdf.hpp

```diff
@@ -0,0 +1,160 @@
+ * log probabilities for each 2-vector observation/correlation
+ * parameter pair.
+ *
+ * @param y (Sequence of) 2-vector(s).
+ * @param rho (Sequence of) correlation parameter(s)
+ * for the standard bivariate normal distribution.
+ * @return The log of the product of the probabilities P(Y1 <= y[1], Y2 <= y[2]
+ * | rho).
+ * @throw std::domain_error if the rho is not between -1 and 1 or nan.
+ * @tparam T_y Underlying type of random variable in sequence.
+ * @tparam T_rho Underlying type of correlation in sequence.
+ */
+
+template <typename T_y, typename T_rho>
+typename return_type<T_y, T_rho>::type std_binormal_lcdf(const T_y& y,
+                                                         const T_rho& rho) {
+  static const char* function = "std_binormal_lcdf";
+  typedef
+      typename stan::partials_return_type<T_y, T_rho>::type T_partials_return;
+  typedef typename stan::math::value_type<T_y>::type T_y_child_type;
+
+  using stan::math::value_of_rec;
+  using std::asin;
+  using std::exp;
+  using std::log;
+  using std::max;
+  using std::sqrt;
+
+  check_bounded(function, "Correlation parameter", rho, -1.0, 1.0);
+  if (stan::is_vector_like<T_y_child_type>::value
+      || stan::is_vector_like<T_rho>::value) {
+    check_consistent_sizes(function, "Random variable", y,
+                           "Correlation parameter", rho);
+  }
+
+  vector_seq_view<T_y> y_vec(y);
+  const scalar_seq_view<T_rho> rho_vec(rho);
+  size_t N_y = length_mvt(y);
+  size_t N_rho = size_of(rho);
+  size_t N = max(N_y, N_rho);
+
+  operands_and_partials<T_y, T_rho> ops_partials(y, rho);
+
+  T_partials_return cdf_log(0.0);
+  if (unlikely(N == 0))
+    return cdf_log;
+
+  // check size consistency of all random variables y
+  for (size_t i = 0; i < N_y; ++i) {
+    check_size_match(function,
+                     "Size of one of the vectors "
+                     "of the location variable",
+                     y_vec[i].size(),
+                     "Size of another vector of "
+                     "the location variable",
+                     2);
+  }
+
+  for (size_t n = 0; n < N_y; ++n) {
+    check_not_nan(function, "Random variable", y_vec[n]);
+  }
+
+  for (size_t n = 0; n < N; ++n) {
+    const T_partials_return y1_dbl = value_of(y_vec[n][0]);
+    const T_partials_return y2_dbl = value_of(y_vec[n][1]);
+    const T_partials_return rho_dbl = value_of(rho_vec[n]);
+
+    T_partials_return cdf_ = std_binormal_integral(y1_dbl, y2_dbl, rho_dbl);
+    cdf_log += log(cdf_);
+
+    if (contains_nonconstant_struct<T_y, T_rho>::value) {
+      const T_partials_return inv_cdf_ = 1.0 / cdf_;
+      const T_partials_return one_minus_rho_sq = (1 + rho_dbl) * (1 - rho_dbl);
+      const T_partials_return sqrt_one_minus_rho_sq = sqrt(one_minus_rho_sq);
+      const T_partials_return rho_times_y2 = rho_dbl * y2_dbl;
+      const T_partials_return y1_minus_rho_times_y2 = y1_dbl - rho_times_y2;
+      const bool y1_isfinite = std::isfinite(value_of_rec(y1_dbl));
```

Comment:
Stan's `is_finite` will work directly on `y1`, I think---I'm not sure why there's a recursive value_of here.

Reply:
I don't think that Stan has an is_finite implementation that works for arbitrary types. Here's an implementation for `vars`: https://github.com/stan-dev/math/blob/502487c511594ccb93eb979df5b8fe163becb417/stan/math/rev/scal/fun/boost_isfinite.hpp but as far as I can tell this is it. I have the `value_of_rec` to handle the nested AD types cases.

Reply:
Thanks for looking.  That's a separate shortcoming of the math lib you don't need to address here.